exercises/go-counting/.meta/src/reference/java/GoCounting.java

```diff
@@ -0,0 +1,157 @@
+import java.awt.Point;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Set;
+
+
+public class GoCounting {
+	int NONE = 0;
+	int WHITE = 1;
+	int BLACK = 2;
+	
+	int[][] board;
+	
+	GoCounting (String boardString) {
+		String[] lines = boardString.split("\n");
+		
+		board = new int[lines[0].length()][lines.length];
+		
+		for(int i = 0; i < lines.length; i++) {
+			for(int j = 0; j < lines[i].length(); j++) {
+				if(lines[i].charAt(j) == 'B') {
+					board[j][i] = BLACK;
+				} else if(lines[i].charAt(j) == 'W') {
+					board[j][i] = WHITE;
+				} else {
+					board[j][i] = NONE;
+				}
+			}
+		}
+	}
+	
+	private ArrayList<Point> getAdjacent(Point p) {
+		ArrayList<Point> adjacent = new ArrayList<>();
+		if(p.x > 0) {
+			adjacent.add(new Point(p.x - 1, p.y));
+		}
+		if(p.x < board.length - 1) {
+			adjacent.add(new Point(p.x + 1, p.y)); 
+		}
+		if(p.y > 0) {
+			adjacent.add(new Point(p.x, p.y - 1));
+		}
+		if(p.y < board[0].length - 1) {
+			adjacent.add(new Point(p.x, p.y + 1));
+		}
+		return adjacent;
+	}
+	
+	int getTerritoryOwner(int x, int y) {
+		
+		if(x < 0 || x >= board.length || y < 0 || y >= board[0].length) {
+			throw new IllegalArgumentException("Invalid coordinate");
+		}
+		
+		if(board[x][y] == BLACK || board[x][y] == WHITE) {
+			return NONE;
+		}
+		
+		ArrayList<Point> visited = new ArrayList<>();
+		ArrayList<Point> edges = new ArrayList<>();
+		ArrayList<Point> territory = new ArrayList<>();
+		
+		ArrayList<Point> tovisit = new ArrayList<>();
+		
+		tovisit.add(new Point(x, y));
+		
+		while(tovisit.size() > 0) {
+			Point current = tovisit.remove(0);
+			
+			if(board[current.x][current.y] == NONE) {
+				visited.add(current);
+				territory.add(current);
+				
+				ArrayList<Point> adjacent = getAdjacent(current);
+				
+				for(Point p : adjacent) {
+					if(!visited.contains(p) && !tovisit.contains(p)) {
+						tovisit.add(p);
+					}
+				}
+			} else {
+				edges.add(current);
+			}
+		}
+		
+		if(edges.size() == 0) {
+			return NONE;
+		}
+		
+		int owner = board[edges.get(0).x][edges.get(0).y];
+		for(int i = 0; i < edges.size(); i++) {
+			if(owner != board[edges.get(i).x][edges.get(i).y]) {
+				owner = NONE;
+			}
+		}
+		
+		return owner;
+	}
+	
+	Set<Point> getTerritory(int x, int y) {
```

Comment:
From what I can tell from the [description of the exercise in problem specifications](https://github.com/exercism/problem-specifications/blob/master/exercises/go-counting/description.md) you only want one method which given a player returns that player's territory. You don't appear to have a function like that from what I can tell? 🙂

Reply:
The getTerritories function does this in a similar manner to how it is done in the Python implementation. I think in C# it is probably similar as well. I based my java implementation on that.

Reply:
Okay, that's fair enough. I think we should try to stick to the description in the readme as much as possible though, especially since there's no canonical data. That way it should be easier for people to understand what they have to do to implement the exercise. So in my opinion we should require the user to implement one function which takes a player (i.e. black or white I think in this game) and returns a set of points.

What do you think @jssander, @sjwarner-bp? 🙂

Reply:
I think that seeing as the PR in problem-specifications contains canonical data and is looking close to being merged, we can review this again with a clearer idea of what we want to achieve as soon as it is merged 🎆